expr/extent/span.go

```diff
@@ -0,0 +1,136 @@
+package extent
+
+import (
+	"fmt"
+
+	"github.com/brimdata/zed/expr"
+	"github.com/brimdata/zed/order"
+	"github.com/brimdata/zed/zng"
+	"github.com/brimdata/zed/zson"
+)
+
+// For now, we do slow-path stuff here but the interface will allow us
+// to optimize with type-specific implementations.  It would be trivial here
+// to create a Time range that embeds nano.Span instead of Lower/Upper
+// and implements Range.
+
+type Span interface {
+	First() zng.Value
+	Last() zng.Value
+	Before(zng.Value) bool
+	After(zng.Value) bool
+	In(zng.Value) bool
+	Overlaps(zng.Value, zng.Value) bool
+	Crop(Span) bool
+	Extend(zng.Value)
+	String() string
+}
+
+type Generic struct {
+	first zng.Value
+	last  zng.Value
+	cmp   expr.ValueCompareFn
+}
+
+// CompareFunc returns a generic comparator suitable for use in a Range
+// based on the order of values in the range, i.e., when order is desc
+// then the first value is larger than the last value and Before is true
+// for larger values while After is true for smaller values, etc.
+func CompareFunc(o order.Which) expr.ValueCompareFn {
+	// Treat nullsmax as zero by passing false to NewValueCompareFn().
+	cmp := expr.NewValueCompareFn(false)
+	//cmp = totalOrderCompare(cmp)
+	if o == order.Asc {
+		return cmp
+	}
+	return func(a, b zng.Value) int { return cmp(b, a) }
+}
+
+// Create a new Range from generic zng.Values.  If first is greater
+// than last according to cmp, then the values are reversed so that
+// first comes first.
+func NewGeneric(first, last zng.Value, cmp expr.ValueCompareFn) *Generic {
```

Comment:
To me a and b is even more confusing.  As I was writing the code I found the convention first/last and lower/upper good pneumonics.  "First" means first wrt to the ordering and "Last" means last wrt to the ordering and this is used consistently throughout the code.  If you try to write all of the methods with a/b, you will quickly get lost in the logic because you will forget when to use < and > on the cmp() results and you will have to reconstruct that Generic.A must be less then Generic.B with respect to Generic.cmp() so better to name them in a way that conveys that invariant in the first place.

Reply:
I used the convention first/last when the order is defined (e.g., in extent.Generic fields) and lower/upper when the order is not defined (e.g., in the ast/dag nodes) so I will use lower/upper here.